### PR TITLE
Integrate snap build script into jenkins job definition

### PR DESCRIPTION
**Pull request by anonymouse64**
This PR generalizes the snap build scripts we have from edgex-go(see [here](https://github.com/edgexfoundry/edgex-go/blob/master/snap/build.sh), [here](https://github.com/edgexfoundry/edgex-go/blob/master/snap/entrypoint.sh), and [here](https://github.com/edgexfoundry/edgex-go/blob/master/snap/Dockerfile.build)) and stores it all in the jenkins job definition directly as the `build_script` so that when we add more repositories, it's easy to create an associated snap job as well by just templating out the job definitions for that project/repo and specifying the `{project-name}-{stream}-stage-snap` and `{project-name}-{stream}-verify-snap`, etc. 

I deployed this job on the sandbox and was able to build the associated edgex-go PR https://github.com/edgexfoundry/edgex-go/pull/1297 successfully. See https://jenkins.edgexfoundry.org/sandbox/view/All/job/edgex-go-snap-master-verify-snap/1/console. I did modify the job definition slightly by removing the macro which adds the snapcraft credentials to the job when running because those credentials aren't available to the sandbox jenkins and they're not really needed to test the job at all. The patch is here if anyone wants to mess around with the sandbox job definition:
```patch
diff --git a/jjb/edgex-templates-snap.yaml b/jjb/edgex-templates-snap.yaml
index 20299d7..b6cee40 100644
--- a/jjb/edgex-templates-snap.yaml
+++ b/jjb/edgex-templates-snap.yaml
@@ -45,7 +45,7 @@
           files:
             - file-id: netrc
               target: '$HOME/.netrc'
-      - edgex-snap-wrapper
+      # - edgex-snap-wrapper
 
     publishers:
       - lf-infra-publish
```

The `edgexfoundry-snapcraft.sh` script we have here is not ideal, as it uses 2 heredocs to basically store 2 other scripts inside the single script so that all of that is available from the job definition directly, but here's the reasoning for why I did that:
* The Dockerfile is okay to store in the script and feed via stdin to the `docker build` process because docker build supports stdin input of the Dockerfile. This doesn't pose any complications or differences in how the docker container is built or runs. We currently need to include all of the files from the project into the docker container as well as the credentials file so that when the build is finished, the snap artifacts don't leave the container and just get pushed into the store when successful. We could use volumes but I was told that this was complicated to setup such that SELinux allowed the mounts to happen and that it was better to just inject all the files we need during the build into the container image. 
* The entrypoint.sh script is a bit more difficult because we need that script to be available to docker as part of it's build context somehow when the container is being built. We can't just store the entrypoint.sh somewhere in /tmp, etc. as it needs to be accessible inside the container. We could have the Dockerfile download the entrypoint.sh script from somewhere (perhaps even this repository), but that leads to complicated syncing problems especially when modifying that file, as you need to update the file in a way that the build when it runs to test actually is able to download that file. As such, the temporary best option is to just store that file inside the build context directly by saving the file into `$GIT_ROOT/snap/entrypoint.sh`.

I also ran into a bit of difficulty with the `edgexfoundry-snapcraft.sh` script and the `entrypoint.sh` heredoc, where the original entrypoint.sh had functions defined for various things and just called those bash functions, but I couldn't figure out how to escape the necessary `{` and `}` characters in such a way that the function definitions were still valid bash and allowed as valid job definition YAML. If I used the `include-raw` macro, it would fail to generate the job definition because it would treat the function definition body as a key to replace with a jenkins job builder variable, so I had to use `{{` and `}}` in the file to generate the appropriate job definition, but then the `{{` weren't re-replaced to be just `{` again and so it wasn't a valid bash script... I also tried numerous ways of escaping the `{` with the `include-raw-escape` macro and everytime the `{` would get replaced into `{{`, so in the end I just removed the functions and called the code directly wherever I needed to use the function. It's not great but it works.

I also setup the delhi override successfully such that we won't need to backport these changes to delhi and delhi jobs can continue to build using the build.sh script present in that repo. You can see that job on the sandbox here: https://jenkins.edgexfoundry.org/sandbox/view/All/job/edgex-go-snap-delhi-stage-snap/1/console (but not the job itself is marked as failure only because it couldn't actually release the snap)

Note that I think we should merge this into ci-management before merging the edgex-go PR so that the edgex-go PR can actually use these job definitions to test those changes.

Finally, I also fixed some typos here and removed the redundant "snap" in the arm verify job name.

### CODE REVIEW

File: shell/edgexfoundry-snapcraft.sh
Comment (eb-oss) [inline shell/edgexfoundry-snapcraft.sh]: *contained

File: shell/edgexfoundry-snapcraft.sh
Comment (eb-oss) [inline shell/edgexfoundry-snapcraft.sh:23]: Does this and the regex two lines below need to include "snap"? If the script is running, can we not assume we are in the correct job type?

File: shell/edgexfoundry-snapcraft.sh
Comment (eb-oss) [inline shell/edgexfoundry-snapcraft.sh:18]: Just want to confirm that we don't want to abort here, but just echo this warning and continue on.

File: shell/edgexfoundry-snapcraft.sh
Comment (eb-oss) [inline shell/edgexfoundry-snapcraft.sh]: *snapcraft.yaml
Also, this is a minor quibble, but the combination of no caps at the beginning of sentences and no periods at the end of some sentences (where EOL seems to be an implicit end of sentence) makes this a bit confusing to read. Readability could certainly be improved.

File: shell/edgexfoundry-snapcraft.sh
Comment (eb-oss) [inline shell/edgexfoundry-snapcraft.sh]: Another sentence that seems to have an implicit end at EOL.

File: shell/edgexfoundry-snapcraft.sh
Comment (eb-oss) [inline shell/edgexfoundry-snapcraft.sh]: *its

**Review by eb-oss — CHANGES_REQUESTED** [review]

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh:18]: The reason not to exit is mainly just for convenience in testing on the sandbox and locally so that I can test locally by running (from the `edgex-go` dir cloned locally):
```bash
JENKINS_URL=something JOB_NAME=edgex-go-snap-master-stage-snap /home/me/git/ci-management/shell/edgexfoundry-snapcraft.sh
```
and have the job continue to run.

We could certainly make it abort here too but then to test on the sandbox and locally I would need to either create a fake file at `$HOME/EdgeX` or modify the script

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh:23]: Yes the regex could leave out snap and still be fine, I assume you mean something like

```bash
    if [[ "$JOB_NAME" =~ .*-stage-snap.* ]]; then
        JOB_TYPE="stage"
    elif [[ "$JOB_NAME" =~ .*-release-snap ]]; then
        JOB_TYPE="release"
    else
        JOB_TYPE="build"
    fi
```

?

Comment (eb-oss) [reply shell/edgexfoundry-snapcraft.sh:23]: I meant if could be just `.*-stage-.*` and `.*-release-.*`

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh]: Fixed (and changed all of the other comments in this file)

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh]: Fixed

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh]: Fixed

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh]: Fixed

File: shell/edgexfoundry-snapcraft.sh
Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh:23]: Ah, yes it could be. Do you want me to change it to use the simpler regex?

Comment (eb-oss) [reply shell/edgexfoundry-snapcraft.sh:23]: To me, simpler means the code is more generalized and adaptable. My preference would be to make it as generic as possible, so long as it won't create false positives.

Comment (anonymouse64) [reply shell/edgexfoundry-snapcraft.sh:23]: Hmm, well I guess after thinking about it some more I would rather leave it at `*-snap-.*-stage-snap.*` and `.*-release-snap` because while yes the script won't ever do anything other than run snap jobs, we may at some point have a repo name like `device-stage` where the job names would then look like `device-stage-snap-release-snap` and `device-stage-snap-master-stage-snap` in which case the simpler regex would cause false positives. I agree with the sentiment of having simpler and generic as possible, but I think stricter is better in this instance.

Comment (eb-oss) [reply shell/edgexfoundry-snapcraft.sh:23]: Ok, that's acceptable to me.

**Review by eb-oss — DISMISSED** [review]
I'd like to see another approval from a dev who is more familiar with the snap building process than I am before merging.

**Comment by anonymouse64** [comment]
@tonyespy can review this when he gets back from Seoul

**Review by jamesrgregg — COMMENTED** [review]
@anonymouse64 I'm not seeing any code in the `ci-snaps-common` repo.  Is there still a plan or a need to make use of that repo?

**Comment by jamesrgregg** [comment]
@tsconn23 please take a look as well since we discussed this today in our Seoul F2F

**Comment by anonymouse64** [comment]
@jamesgregg that repository, if needed, will be referenced from the snapcraft.yaml files in individual repositories, not from here directly. It's also not clear to me if I will need that repository at all as I didn't end up needing it for edgex-go or some of the other snaps I've prototyped with.

**Comment by anonymouse64** [comment]
I modified the Dockerfile slightly to change how we "install" snaps there to be less repetitive and also to add the core18 snap there too, which while not strictly necessary is useful for snapcraft in determining missing library dependencies. I re-pushed the job to the sandbox here: https://jenkins.edgexfoundry.org/sandbox/view/All/job/edgex-go-snap-master-verify-snap/6/console and the build seems to be working (it hasn't yet finished however).

**Review by eb-oss — DISMISSED** [review]

File: shell/edgexfoundry-snapcraft.sh
Comment (tonyespy) [reply shell/edgexfoundry-snapcraft.sh]: Did you fix by adding the apostrophe? The current usage is incorrect.

File: shell/edgexfoundry-snapcraft.sh
Comment (tonyespy) [inline shell/edgexfoundry-snapcraft.sh]: snapcraf.yaml --> snapcraft.yaml

**Review by tonyespy — COMMENTED** [review]
One general comment/nit I have is that your commit message summaries all include a path to the modified file as a prefix which causes the line length to be >50 chars. Out of 300+ commits in this repo, only five commits use this kind of prefix...

Otherwise the approach and implementation in this PR seems correct to me.

Please see my two other inline comments.

And finally, looks like a pesky "merge master into <feature branch>" was introduced recently...

**Comment by eb-oss** [comment]
I agree with Tony that the first line of the commit should be limited to 50 chars. You can expand on that in the description, but I also tend to think that putting the path in the commit message is unnecessary, as the commit itself holds file change info.

That "merge master into..." was due to me using the "update branch" button on this PR. I was thinking it was a rebase (as it is in Gerrit, which LF uses much more extensively than Github), so I apologize for that. Not the end of the world to have in the history, but I agree that it's annoying.

**Comment by anonymouse64** [comment]
I adjusted the commit messages. FWIW, I think that looking back on history seeing the file in the summary is very useful because many times the commit makes a specific change to a specific file and referring to the change alone isn't specific enough - a similar change could have (or in some cases already has) been made in a different file from the one that the commit is making and you want to be able to quickly tell what a commit changed. Additionally, the file that was changed allows you to quickly skim down a list of commit summaries and see what files they changed - in projects that follow this pattern I can go back to the project after a lot of work has been done and just by looking at the commit summaries with their filenames see which commits might be of interest to me that I should look at. It's not a replacement for `git blame`, but a good way to skim over git history.

**Comment by anonymouse64** [comment]
I also think the whole 50 chars vs 72 chars is arbitrary...

**Review by tonyespy — DISMISSED** [review]
LGTM

**Review by eb-oss — DISMISSED** [review]
LGTM. @jamesrgregg should I wait for @tsconn23 review before merging?

**Comment by tsconn23** [comment]
@MightyNerdEric Don't wait for me given the approvals above.

**Review by eb-oss — APPROVED** [review]

**Pull request merged**
